csclient/csclient.go

```diff
@@ -0,0 +1,487 @@
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+	resp.Body.Close()
+	return nil
+}
+
+// Meta fetches metadata on the charm or bundle with the
+// given id. The result value provides a value
+// to be filled in with the result, which must be
+// a pointer to a struct containing members corresponding
+// to possible metadata include parameters (see http://tinyurl.com/nysdjly).
+//
+// It returns the fully qualified id of the entity.
+//
+// The name of the struct member is translated to
+// a lower case hyphen-separated form; for example,
+// ArchiveSize becomes "archive-size", and BundleMachineCount
+// becomes "bundle-machine-count", but may also
+// be specified in the field's tag
+//
+// This example will fill in the result structure with information
+// about the given id, including information on its archive
+// size (include archive-size), upload time (include archive-upload-time)
+// and digest (include extra-info/digest).
+//
+//	var result struct {
+//		ArchiveSize params.ArchiveSizeResponse
+//		ArchiveUploadTime params.ArchiveUploadTimeResponse
+//		Digest string `csclient:"extra-info/digest"`
+//	}
+//	id, err := client.Meta(id, &result)
+func (c *Client) Meta(id *charm.Reference, result interface{}) (*charm.Reference, error) {
+	if result == nil {
+		return nil, fmt.Errorf("expected valid result pointer, not nil")
+	}
+	resultv := reflect.ValueOf(result)
+	resultt := resultv.Type()
+	if resultt.Kind() != reflect.Ptr {
+		return nil, fmt.Errorf("expected pointer, not %T", result)
+	}
+	resultt = resultt.Elem()
+	if resultt.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("expected pointer to struct, not %T", result)
+	}
+	resultv = resultv.Elem()
+
+	// At this point, resultv refers to the struct value pointed
+	// to by result, and resultt is its type.
+
+	numField := resultt.NumField()
+	includes := make([]string, 0, numField)
+
+	// results holds an entry for each field in the result value,
+	// pointing to the value for that field.
+	results := make(map[string]reflect.Value)
+	for i := 0; i < numField; i++ {
+		field := resultt.Field(i)
+		if field.PkgPath != "" {
+			// Field is private; ignore it.
+			continue
+		}
+		if field.Anonymous {
+			// At some point in the future, it might be nice to
+			// support anonymous fields, but for now the
+			// additional complexity doesn't seem worth it.
+			return nil, fmt.Errorf("anonymous fields not supported")
+		}
+		apiName := field.Tag.Get("csclient")
+		if apiName == "" {
+			apiName = hyphenate(field.Name)
+		}
+		includes = append(includes, "include="+apiName)
+		results[apiName] = resultv.FieldByName(field.Name).Addr()
+	}
+	// We unmarshal into rawResult, then unmarshal each field
+	// separately into its place in the final result value.
+	// Note that we can't use params.MetaAnyResponse because
+	// that will unpack all the values inside the Meta field,
+	// but we want to keep them raw so that we can unmarshal
+	// them ourselves.
+	var rawResult struct {
```

Comment:
For the record, with httprequest this might look something like:

	type metaAnyRequest struct {
		Include []string `httprequest:"include,form"`
		Result struct {
			Id *charm.Reference
			Meta map[string] json.RawMessage
		}		`httprequest:",body"`,
	}